src/rdmap.c

```diff
@@ -0,0 +1,487 @@
+        if (r != 0)
+                return r;
+        return rmap->rmap_cmp(a->key, b->key);
+}
+
+static void rd_map_elem_destroy (rd_map_t *rmap, rd_map_elem_t *elem) {
+        rd_assert(rmap->rmap_cnt > 0);
+        rmap->rmap_cnt--;
+        if (rmap->rmap_destroy_key)
+                rmap->rmap_destroy_key((void *)elem->key);
+        if (rmap->rmap_destroy_value)
+                rmap->rmap_destroy_value((void *)elem->value);
+        LIST_REMOVE(elem, hlink);
+        LIST_REMOVE(elem, link);
+        rd_free(elem);
+}
+
+static rd_map_elem_t *rd_map_find (const rd_map_t *rmap, int *bktp,
+                                   const rd_map_elem_t *skel) {
+        int bkt = skel->hash % rmap->rmap_buckets.cnt;
+        rd_map_elem_t *elem;
+
+        if (bktp)
+                *bktp = bkt;
+
+        LIST_FOREACH(elem, &rmap->rmap_buckets.p[bkt], hlink) {
+                if (!rd_map_elem_cmp(skel, elem, rmap))
+                        return elem;
+        }
+
+        return NULL;
+}
+
+
+/**
+ * @brief Create and return new element based on \p skel without value set.
+ */
+static rd_map_elem_t *rd_map_insert (rd_map_t *rmap, int bkt,
+                                     const rd_map_elem_t *skel) {
+        rd_map_elem_t *elem;
+
+        elem = rd_calloc(1, sizeof(*elem));
+        elem->hash = skel->hash;
+        elem->key = skel->key; /* takes ownership of key */
+        LIST_INSERT_HEAD(&rmap->rmap_buckets.p[bkt], elem, hlink);
+        LIST_INSERT_HEAD(&rmap->rmap_iter, elem, link);
+        rmap->rmap_cnt++;
+
+        return elem;
+}
+
+
+void rd_map_set (rd_map_t *rmap, void *key, void *value) {
+        rd_map_elem_t skel = { .key = key,
+                               .hash = rmap->rmap_hash(key) };
+        rd_map_elem_t *elem;
+        int bkt;
+
+        if (!(elem = rd_map_find(rmap, &bkt, &skel))) {
+                elem = rd_map_insert(rmap, bkt, &skel);
+        } else {
+                if (elem->value && rmap->rmap_destroy_value)
+                        rmap->rmap_destroy_value((void *)elem->value);
+                if (rmap->rmap_destroy_key)
+                        rmap->rmap_destroy_key(key);
+        }
+
+        elem->value = value; /* takes ownership of value */
+}
+
+
+void *rd_map_get (const rd_map_t *rmap, const void *key) {
+        const rd_map_elem_t skel = { .key = (void *)key,
+                                     .hash = rmap->rmap_hash(key) };
+        rd_map_elem_t *elem;
+
+        if (!(elem = rd_map_find(rmap, NULL, &skel)))
+                return NULL;
+
+        return (void *)elem->value;
+}
+
+
+rd_map_elem_t *rd_map_get_or_create (rd_map_t *rmap, const void *key) {
+        const rd_map_elem_t skel = { .key = (void *)key,
+                                     .hash = rmap->rmap_hash(key) };
+        rd_map_elem_t *elem;
+        int bkt;
+
+        if ((elem = rd_map_find(rmap, &bkt, &skel)))
+                return elem;
+
+        return rd_map_insert(rmap, bkt, &skel);
```

Comment:
this takes ownership of `key`, but `rd_map_find` does not, so the caller of `rd_map_get_or_create` can't know if they need to free `key` or not, right?

Reply:
Good point, since this method isn't used I'll remove it.